src/engine/strat_engine/thinpool.rs

```diff
@@ -73,46 +92,89 @@ impl ThinPool {
+                                           new_meta_dev,
+                                           data_dev)
+                                .map(|dev| (dev, new_spare_segments))
+                                .map_err(|e| e.into())
+                    })
             }
-            Err(DmError::Dm(devicemapper::ErrorEnum::CheckFailed(meta_dev, data_dev), _)) => {
-                let (new_meta_dev, new_spare_segments) =
-                    try!(attempt_thin_repair(pool_uuid, dm, meta_dev, spare_segments));
-                Ok(ThinPool {
-                       thin_pool: try!(ThinPoolDev::setup(&name,
-                                                          dm,
-                                                          size,
-                                                          data_block_size,
-                                                          low_water_mark,
-                                                          new_meta_dev,
-                                                          data_dev)),
-                       meta_spare: new_spare_segments,
-                       id_gen: ThinDevIdPool::new_from_ids(thin_ids),
-                   })
+            Err(e) => Err(e.into()),
+        };
+        let (thinpool_dev, spare_segments) = try!(res);
+
+        let filesystems: Vec<StratFilesystem> = {
+            /// Set up a filesystem from its metadata.
+            let get_filesystem = |fssave: &FilesystemSave| -> EngineResult<StratFilesystem> {
+                Ok(try!(StratFilesystem::setup(pool_uuid,
+                                               fssave.uuid,
+                                               fssave.thin_id,
+                                               &fssave.name,
+                                               fssave.size,
+                                               dm,
+                                               &thinpool_dev)))
+            };
+
+            try!(fs_save.iter().map(get_filesystem).collect())
+        };
+
+        let mut fs_table = Table::default();
+        for fs in filesystems {
+            let evicted = fs_table.insert(fs);
+            if !evicted.is_empty() {
+                let err_msg = "filesystems with duplicate UUID or name specified in metadata";
+                return Err(EngineError::Engine(ErrorEnum::Invalid, err_msg.into()));
             }
-            Err(err) => Err(err.into()),
         }
+
+        Ok(ThinPool {
+               thin_pool: thinpool_dev,
+               meta_spare: spare_segments,
+               id_gen: ThinDevIdPool::new_from_ids(thin_ids),
+               filesystems: fs_table,
+               mdv: mdv,
+           })
     }
 
     /// The status of the thin pool as calculated by DM.
     pub fn thin_pool_status(&self, dm: &DM) -> EngineResult<ThinPoolStatus> {
-        Ok(try!(self.thin_pool.status(dm)))
+        try!(self.thin_pool.status(dm));
+        try!(self.mdv.check());
+
+        for fs in &self.filesystems {
```

Comment:
The only value this method can return is ```ThinPoolStatus::Good``` or an error of some kind. That makes the entire match starting at line 328 useless, AFAICT.